File: Sean的leetcode/FB/Inorder Successor in BST.py
###
# BST
# Time Complexity: O(h)
# Space Complexity: O(1)
###
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderSuccessor(self, root, p):
        """
        :type root: TreeNode
        :type p: TreeNode
        :rtype: TreeNode
        """
        if not root or not p:
            return None
        succ = None
        cur = root
        while cur:
            if cur.val > p.val:
                succ = cur
                cur = cur.left
            else:
                cur = cur.right
        return succ

File: Sean的leetcode/Leetcode/ Inorder Successor in BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
//Time: O(h), h is tree height.
class Solution {
public:
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        if (!root || !p) return 0;

        TreeNode *curr = root, *succ;
        while (curr)
        {
            if (curr->val == p->val)
                curr = curr->right;
            else if (curr->val > p->val)
            {
                succ = curr;
                curr = curr->left;
            }
            else
                curr = curr->right;
        }
        return succ;
    }
};